ipa-core/src/protocol/basics/check_zero.rs

```diff
@@ -124,4 +187,122 @@ mod tests {
+        pub fn get(&self) -> Pin<Box<dyn Future<Output = T> + Send + '_>> {
+            let inner = self.inner.lock().unwrap();
+            if let Some(value) = inner.cell.get() {
+                return ready(value.clone()).boxed();
+            }
+            let notify = inner.notify.clone();
+            async move {
+                notify.notified().await;
+                self.inner.lock().unwrap().cell.get().unwrap().clone()
+            }
+            .boxed()
+        }
+    }
+
+    struct MaliciousCheckZeroInterceptor {
+        sh2: NotifyOnceCell<Fp32BitPrime>,
+    }
+
+    impl MaliciousCheckZeroInterceptor {
+        fn new() -> Self {
+            Self {
+                sh2: NotifyOnceCell::new(),
+            }
+        }
+    }
+
+    impl StreamInterceptor for MaliciousCheckZeroInterceptor {
+        type Context = InspectContext;
+
+        fn peek<'a>(
+            &'a self,
+            ctx: &'a Self::Context,
+            data: &'a mut Vec<u8>,
+        ) -> Pin<Box<dyn Future<Output = ()> + Send + 'a>> {
+            if ctx
+                .gate
+                .as_ref()
+                .contains(super::Step::MultiplyWithR.as_ref())
+                && ctx.identity == HelperIdentity::ONE
+                || ctx.gate.as_ref().contains(super::Step::RevealR.as_ref())
+                    && ctx.identity == HelperIdentity::ONE
+                    && ctx.dest == HelperIdentity::TWO.as_str()
+            {
+                async {
+                    assert_eq!(data.len(), 4);
+                    let (sh1, sh2) = join(super::SH1.get(), self.sh2.get()).await;
+                    tracing::info!("got shares: {sh1:?} {sh2:?}");
+                    let adjusted_share = -sh1 - sh2;
+                    tracing::info!("adjusted share {adjusted_share:?}");
+                    adjusted_share.serialize(
+                        <&mut GenericArray<u8, U4>>::try_from(data.as_mut_slice()).unwrap(),
+                    );
+                }
+                .boxed()
+            } else if ctx.gate.as_ref().contains(super::Step::RevealR.as_ref())
+                && ctx.identity == HelperIdentity::TWO
+                && ctx.dest == HelperIdentity::ONE.as_str()
+            {
+                assert_eq!(data.len(), 4);
+                let sh2 = Fp32BitPrime::deserialize_unchecked(
+                    <&GenericArray<u8, U4>>::try_from(data.as_slice()).unwrap(),
+                );
+                self.sh2.set(sh2).unwrap();
+                tracing::info!("sent sh2 = {sh2:?}");
+                ready(()).boxed()
+            } else {
+                ready(()).boxed()
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn malicious_check_zero() {
+        let mut config = TestWorldConfig::default();
+        config.stream_interceptor = Arc::new(MaliciousCheckZeroInterceptor::new());
+        let world = TestWorld::new_with(&config);
+        let mut rng = thread_rng();
+        let v = rng.gen::<Fp32BitPrime>();
```

Comment:
Does this test benefit from using a random value or we can use a constant here to avoid false negatives?